src/types/mod.rs

```diff
@@ -1,160 +1,219 @@
 use itertools::Itertools;
-use std::cmp::Ordering;
 use std::collections::{HashMap, HashSet};
-use std::hash::{Hash, Hasher};
 use std::iter;
 
 use super::parser;
 
-/// Represents a unique identity, a [DeBruijn Index].
-///
-/// [DeBruijn Index]: https://en.wikipedia.org/wiki/De_Bruijn_index
-pub type DeBruijn = usize;
-
-/// Represents a place in a [`Term`]
+/// Represents a place in a [`Term`].
 ///
 /// [`Term`]: struct.Term.html
-pub type Place = Vec<usize>;
-
-/// An ADT for Variables.
-pub trait Variable: Eq + Ord + Hash + Clone {
-    /// Construct a new `Variable` that has not been used.
-    fn new_distinct<'a, T>(existing: T, name: Option<String>) -> Self
-    where
-        T: IntoIterator<Item = &'a Self>,
-        Self: 'a;
-    /// A human-readable representation of the variable.
-    fn show(&self) -> String {
-        "<variable>".to_string()
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Place(Vec<usize>);
+impl Place {
+    pub fn location(&self) -> &[usize] {
+        &self.0
     }
 }
 
-/// An ADT for Operators.
-pub trait Operator: Eq + Hash + Clone {
-    /// Construct a new `Operator` that has not been used.
-    fn new_distinct<'a, T>(operators: T, arity: usize, name: Option<String>) -> Self
-    where
-        T: IntoIterator<Item = &'a Self>,
-        Self: 'a;
-    /// The number of arguments a particular operator takes.
-    fn arity(&self) -> usize;
-    /// A human-readable representation of the operator.
-    fn show(&self) -> String {
-        "<operator>".to_string()
-    }
+/// A symbol for an unspecified term.
+///
+/// To construct an [`Variable`], use [`Signature::new_var`]
+///
+/// [`Variable`]: struct.Variable.html
+/// [`Signature::new_var`]: struct.Signature.html#method.new_var
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Variable {
```

Comment:
This is something the changed a lot between my iterations of implementation for this. Basically, it cannot be done in an elegant way. Signatures should be mutable, so that new variables and operators can be constructed. Using `RefCell` would make this possible, but would add enormous complexity.